sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -163,6 +171,21 @@ static BundleMethod create(Method targetMethod) {
     }
   }
 
+  /**
+   * Describes a state declaration; a field of type {@link StateSpec} annotated with
+   * {@link DoFn.StateId}.
+   */
+  @AutoValue
+  public abstract static class StateDeclaration {
+    public abstract String id();
+    public abstract Field field();
+    public abstract TypeDescriptor<?> stateType();
```

Comment:
Done. It is statically assured in the analysis, but this ensures the knowledge from the analysis is maintained in the data structure.